cdc/puller/unified_sorter.go

```diff
@@ -0,0 +1,686 @@
+	maxResolvedTs uint64
+	finished      chan error
+	dealloc       func() error
+}
+
+type heapSorter struct {
+	id          int
+	inputCh     chan *model.PolymorphicEvent
+	outputCh    chan *flushTask
+	heap        sortHeap
+	backEndPool *backEndPool
+}
+
+func newHeapSorter(id int, pool *backEndPool, out chan *flushTask) *heapSorter {
+	return &heapSorter{
+		id:          id,
+		inputCh:     make(chan *model.PolymorphicEvent, 1024*1024),
+		outputCh:    out,
+		heap:        make(sortHeap, 0, 65536),
+		backEndPool: pool,
+	}
+}
+
+// flush should only be called within the main loop in run().
+func (h *heapSorter) flush(ctx context.Context, maxResolvedTs uint64) error {
+	isEmptyFlush := h.heap.Len() == 0
+	var backEnd sorterBackEnd = nil
+
+	if !isEmptyFlush {
+		var err error
+		backEnd, err = h.backEndPool.alloc()
+		if err != nil {
+			return errors.AddStack(err)
+		}
+	}
+
+	task := &flushTask{
+		heapSorterId:  h.id,
+		backend:       backEnd,
+		maxResolvedTs: maxResolvedTs,
+		finished:      make(chan error),
+	}
+
+	var oldHeap sortHeap
+	if !isEmptyFlush {
+		task.dealloc = func() error {
+			return h.backEndPool.dealloc(backEnd)
+		}
+		oldHeap = h.heap
+		h.heap = make(sortHeap, 0, 65536)
+	} else {
+		task.dealloc = func() error {
+			return nil
+		}
+	}
+
+	log.Debug("Unified Sorter new flushTask", zap.Int("heap-id", task.heapSorterId),
+		zap.Uint64("resolvedTs", task.maxResolvedTs))
+	go func() {
+		defer close(task.finished)
+		if isEmptyFlush {
+			return
+		}
+		batchSize := oldHeap.Len()
+		for oldHeap.Len() > 0 {
+			event := heap.Pop(&oldHeap).(*sortItem).entry
+			err := task.backend.writeNext(event)
+			if err != nil {
+				task.finished <- err
+				return
+			}
+		}
+		err := task.backend.flush()
+		if err != nil {
+			task.finished <- err
+		}
+
+		log.Debug("Unified Sorter flushTask finished",
+			zap.Int("heap-id", task.heapSorterId),
+			zap.Uint64("resolvedTs", task.maxResolvedTs),
+			zap.Int("size", batchSize))
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case h.outputCh <- task:
+	}
+	return nil
+}
+
+func (h *heapSorter) run(ctx context.Context) error {
+	var (
+		maxResolved           uint64
+		heapSizeBytesEstimate int64
+	)
+	maxResolved = 0
+	heapSizeBytesEstimate = 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case event := <-h.inputCh:
+			heap.Push(&h.heap, &sortItem{entry: event})
+			isResolvedEvent := event.RawKV != nil && event.RawKV.OpType == model.OpTypeResolved
+			if isResolvedEvent {
+				log.Debug("heapSorter got resolved event", zap.Uint64("CRTs", event.RawKV.CRTs), zap.Int("heap-id", h.id))
+				if event.RawKV.CRTs < maxResolved {
+					log.Fatal("ResolvedTs regression, bug?", zap.Uint64("event-resolvedTs", event.RawKV.CRTs),
+						zap.Uint64("max-resolvedTs", maxResolved))
+				}
+				maxResolved = event.RawKV.CRTs
+			}
+
+			heapSizeBytesEstimate += event.RawKV.ApproximateSize()
+			if heapSizeBytesEstimate >= heapSizeLimit || isResolvedEvent {
```

Comment:
Maybe we need to calculate the estimated size before adding an entry to the heap to make sure it does not exceed the limit.